drivers/firmware/rp1.c

```diff
@@ -270,19 +270,18 @@ static int rp1_firmware_probe(struct platform_device *pdev)
 	init_completion(&fw->c);
 	kref_init(&fw->consumers);
 
-	platform_set_drvdata(pdev, fw);
-
 	ret = rp1_firmware_message(fw, GET_FIRMWARE_VERSION,
 				   NULL, 0, &version, sizeof(version));
 	if (ret == sizeof(version)) {
 		dev_info(dev, "RP1 Firmware version %08x%08x%08x%08x%08x\n",
 			 version[0], version[1], version[2], version[3], version[4]);
-		ret = 0;
-	} else if (ret >= 0) {
-		ret = -EIO;
+		platform_set_drvdata(pdev, fw);
+	} else {
+		kfree(fw);
+		platform_set_drvdata(pdev, ERR_PTR(-ENOENT));
```

Comment:
We do dereference this in `rp1_firmware_remove` (without a check). 
I'm not sure if that can happen, but looks like a potential pitfall.

Reply:
Testing the obvious fix has uncovered some deficiencies in the end-of-life code. Thank goodness for runtime overlays.